### ISSUE TITLE
Create a public C++ API

### ISSUE THREAD

**Issue by jbigot** [pdidev/pdi#131]
This is the head issue related to the proposition of a new C++ API.

The main idea is as follow.

`PDI::Data_Store`:
* the core of PDI should be a `PDI::Data_Store` class,
* an instance of `Data_Store` is accessible from `PDI::Context` by a `data()` member function,
* the `Data_Store` class is very similar to a `std::unordered_map<std::string, PDI::Weak_Ref>`,
* in addition, it is "observable", i.e. one can register to be notified when references are added/removed in it.

`PDI::Weak_Ref`/`PDI::Shared_Ref`/`PDI::Shared_R_Ref`/`PDI::Shared_RW_Ref`
* the `PDI::Shared_Ref`/`PDI::Weak_Ref` classes can be used outside of PDI, they behave similarly to `std::shared_ptr`/`std::weak_ptr`,
  - `PDI::Weak_Ref`/`PDI::Shared_Ref` can not be directly dereferenced,
  - when dereferenced, `PDI::Shared_R_Ref` returns a const pointer,
  - when dereferenced, `PDI::Shared_R_Ref` returns a normal pointer,
* these references implement locking:
  - `PDI::Shared_R_Ref` behaves like a `std::shared_lock`,
  - `PDI::Shared_W_Ref` behaves like a `std::unique_lock`,
* the memory ownership of `shared_ref` is weaker than that of `shared_ptr` (see Java `SoftReference`):
  - all reference types offer a `release()` member function similarly to `std::unique_ptr`,
  - the `reclaim()` variant release all reference to the exception of the current one,
  - all reference types offer a callback mechanism to be notified before the referenced data is released,
* unlike their C++ counterpart, these references offer dynamic typing
  - a `type()` member function is offered that returns a `Datatype` to retrieve the type information,
  - templated versions are offered so that the user doesn't have to cast the `void*` pointer on every use,
* **TODO**: add a word about metadata & type template system

`g_user_ref`:
* the `unordered_map<string, std::unique_ptr<Ref_holder>> g_user_ref` global variable stores the user-side references of the C API users,
* `Ref_holder` is a base class that is specialized with versions holding any of `PDI::Weak_Ref`/`PDI::Shared_Ref`/`PDI::Shared_R_Ref`/`PDI::Shared_RW_Ref`,
* `Ref_holder` can always be converted to a `PDI::Weak_Ref`.

The C API behaviour is as follow:
* `share` creates a new reference and stores it both in the global `Context` and in `g_user_ref`,
* `access` gets a reference from the global `Context` and store it in `g_user_ref`,
* `release` removes the reference from `g_user_ref`,
* `reclaim` calls `release()` on the reference from `g_user_ref` and removes it.

In other words
```C++
void* my_data = //...

// this
PDI_share("name", ptr, PDI_INOUT);
// is equivalent to this
Shared_RW_Ref my_ref(ptr);
pdi.data["name"] = my_ref;

// and this
PDI_release("name");
// is equivalent to this
my_ref.~Shared_RW_Ref(); //< usually done automatically
```
and
```C++
// this
PDI_access("name", *ptr, PDI_INOUT);
// is equivalent to this
Shared_RW_Ref my_ref = pdi.data["name"];

// and this
PDI_reclaim("name");
// is equivalent to this
my_ref.reclaim();
// or this
ptr = my_ref.release();
```

**Comment by jbigot** [comment]
The initial proposition by @ksiero and @tpaluszk was as follow:

Credit where credit's due, this is work by @ksiero and @tpaluszk

`PDI`: This is the class that user will use to initialize PDI. It holds `PDI_Core` and references to the `data` and `events` to the `PDI_Core`. User can add new datatypes and plugins (in case of non-config run). Also we give him a custom logger.

`PDI_core`: This is our container for `Event_map`, `Data_root` and `Plugins` and `init_callbacks`.

`Event_map`: Holds `any_event_callbacks` (callbacks called on any event) and maps of individual `Event`s.
`Event`: Simple class that holds "named" event callbacks and calls all callbacks from `Event_map::m_any_event_callbacks`.

`Data_root`: This is class for user to access the data nodes, also user can add custom callbacks on:
1. `Share` - data becoming available
2. `Empty_access` - user tries to access node that doesn't have a shared pointer
3. `Nulify` - when shared pointer on this node is deleted

`Data_root_for_nodes`: gives access to `any_*_callbacks` list for all nodes.

`Node`: This is the old "Descriptor". We can call:
1. `share` - this creates a `Semaphore` pointer with no access rights, calls all share callbacks 
2. `share_read` - this creates a `Semaphore` pointer with read access
2. `share_write` - this creates a `Semaphore` pointer with write access
3. `share_read_write` - this creates a `Semaphore` pointer with read and write access
   * all these methods have option to pass also a deleter in case of `release()`
4. `access_read` - returns `Read_sptr` which gives read access to the pointer
5. `access_write` - returns `Write_sptr` which gives write access to the pointer
6. `access_read_write` - returns Read_write_sptr which gives read and write access to the pointer
7. `expose`, `expose_read`, `expose_write`, `expose_read_write` - the same as share (but at the end the reclaim is called)
8. `readable()` - true if `m_data_sem_ptr` is set and can create `Read_sptr`
9. `writable()` - true if `m_data_sem_ptr` is set and can create `Write_sptr`
10. `empty()` - `true` if no data is shared for this node
11. cast overload - in case of scalar we can get value from pointer (if there is read access) and return an integer (e.g. int x = PDI.data["x"];)
12. `operator[](size_t)` - used mostly for arrays (e.g. int x = PDI.data["array_of_ints"][0])
13. `operator[](std::string)` - used to access the child and in case of record to access to member
14. `add_*_callbacks` - adds new callback

`Semaphore_ptr` - similar to `Ref_base::Referenced`. It holds the pointer to data and has mutexes and read locks.

`Read_sptr`, `Write_sptr`, `Read_write_sptr` - it is created when someone (plugin or user) want to get access to the pointer. It increments read_locks (in case of read) and lock write mutex (in case of write). If it can't get wanted access throws an exception.

![PDI_UML](https://gitlab.maisondelasimulation.fr/pdidev/pdi/uploads/8b10d49971b944c58f0d24f6e583fd5c/PDI_UML.png)

Share example:
```C++
#include <pdi.h>
#include <paraconf.h>

const char* CONFIG_YAML =
    "data:                         \n"
    "  x: int64                    \n"
    "  array:                      \n"
	"    type: array               \n"
	"    subtype: int              \n"
	"    size: 32                  \n"
    "  record:                     \n"
    "    type: record              \n"
	"    buffersize: 8             \n"
    "    members:                  \n"
    "      a:                      \n"
    "        disp: 0               \n"
    "        type: int             \n"
	"      b:                      \n"
    "        disp: 4               \n"
    "        type: int             \n"
    "plugins:                      \n"
    "  decl_hdf5:                  \n"
    "    file: output.h5           \n"
    "    write: [ x, array, record ]\n";

struct Record {
	int a;
	int b;
};

int main() {
	long* x = new long;
	*x = 0;
	int array[32] = {0};
	Record rec{0, 1};
	PDI::PDI pdi{PC_parse_string(CONFIG_YAML)};
	pdi.data["x"].expose_read(x, [](void* x){delete static_cast<long*>(x);});
	pdi.data["array"].expose_read(array);
	pdi.data["record"].expose_read(&rec);
	pdi.event["end_of_main"].call(); //some event
	return 0;
}
```


Share access:
```C++
#include <iostream>

#include <pdi.h>
#include <paraconf.h>

const char* CONFIG_YAML =
    "data:                         \n"
    "  x: int64                    \n"
    "  array:                      \n"
	"    type: array               \n"
	"    subtype: int              \n"
	"    size: 32                  \n"
    "  record:                     \n"
    "    type: record              \n"
	"    buffersize: 8             \n"
    "    members:                  \n"
    "      a:                      \n"
    "        disp: 0               \n"
    "        type: int             \n"
	"      b:                      \n"
    "        disp: 4               \n"
    "        type: int             \n";

struct Record {
	int a;
	int b;
};

long sum_all_shared_values(PDI::PDI& pdi) {
	long x = 0;
	PDI::Read_sptr array_sptr;
	PDI::Read_sptr rec_sptr;
	try {
		x = pdi.data["x"]; // int cast operator overload (makes a copy)
		array_sptr = pdi.data["array"].access_read();
		rec_sptr = pdi.data["record"].access_read();
	} catch (PDI::Error e) {
		std::cerr << e.what() << std::endl;
		return -1;
	}
	long sum = x;
	for (int i = 0; i < 32; i++) {
		sum += *(array_sptr.get<const int*>()) + i);
	}
	sum += rec_sptr.get<const Record*>()->a + rec_sptr.get<const Record*>()->b;
	return sum;
}

int main() {
	long* x = new long;
	*x = 0;
	int array[32] = {1};
	Record rec{2, 3};
	PDI::PDI pdi{PC_parse_string(CONFIG_YAML)};
	pdi.data["x"].share_read(x, [](void* x){delete static_cast<long*>(x);});
	pdi.data["array"].share_read(array);
	pdi.data["record"].share_read(&rec);

	int shared_sum = sum_all_shared_values(pdi);

	pdi.data["x"].release();
	pdi.data["array"].reclaim();
	pdi.data["record"].reclaim();

	return 0;
}
```

**Comment by jbigot** [comment]
In GitLab by @ksiero on Jan 22, 2020, 15:48

I was thinking about the new C++ API. I came up with `pdi_ptr`, it's something like `shared_ptr`, but with R/W rights. Please see the use of this idea in the attachment: [pdi_cpp_API.cpp](https://gitlab.maisondelasimulation.fr/pdidev/pdi/uploads/bf9a533636bfc2809e6d842e38f704d9/pdi_cpp_API.cpp)
